ci: add workflow for code check

gwen-lg opened the pull request:
Check fmt, lints (clippy), doc & test.

gwen-lg committed:
clippy: replace use of `get` + `is_none` with `contains_key`

warning: unnecessary use of `get(&ElementId::Cluster).is_none()`
         if seek_head.get(&ElementId::Cluster).is_none() {
            help: replace it with: `!seek_head.contains_key(&ElementId::Cluster)`

help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#unnecessary_get_then_check
note: `#[warn(clippy::unnecessary_get_then_check)]` on by default

gwen-lg committed:
clippy: remove redundant guard in matching

warn(clippy::redundant_guards)
help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#redundant_guards

gwen-lg committed:
ci: add workflow for code check

Check fmt, lints (clippy), doc & test.

hasenbanck commented:
Thank you for the CI implementation. I will merge this PR last.

gwen-lg commented:
With interest :smile: .
For a tool written in Rust (in development), I use matroska-demuxer to scan tracks, and call mkvextract in command line.
But I would like to be able to process content (subtitle track in this case) directly from rust code.
Would you be interested to add functionality to be able to read tracks content of mkv in `matroska-demuxer` ?

gwen-lg commented:
I forgot to setup as draft.

Currently, the CI fail.
Clippy as default failed on errors fixed by to first commit of lint PR.
Fixed by the two commits : ([clippy: replace use of get + is_none with contains_key](https://github.com/hasenbanck/matroska-demuxer/pull/12/commits/b03dc4487bb3cea729573effbee31aef6a40abcc) and [clippy: remove redundant guard in matching](https://github.com/hasenbanck/matroska-demuxer/pull/12/commits/ea11482037a532d39c1644b4a4bcea6e4a85cc31)).
Fixes should be integrated or lints disabled for the CI to pass.

A lot of test failed because of missing files, you can take a look at logs :
https://github.com/gwen-lg/matroska-demuxer/actions/runs/13187211187/job/36812072574

hasenbanck commented:
> Currently, the CI fail. Clippy as default failed on errors fixed by to first commit of lint PR. Fixed by the two commits : ([clippy: replace use of get + is_none with contains_key](https://github.com/hasenbanck/matroska-demuxer/pull/12/commits/b03dc4487bb3cea729573effbee31aef6a40abcc) and [clippy: remove redundant guard in matching](https://github.com/hasenbanck/matroska-demuxer/pull/12/commits/ea11482037a532d39c1644b4a4bcea6e4a85cc31)). Fixes should be integrated or lints disabled for the CI to pass.

Then the fixed for the standard lints should be included in this branch as an extra acommit.
 
> A lot of test failed because of missing files, you can take a look at logs : https://github.com/gwen-lg/matroska-demuxer/actions/runs/13187211187/job/36812072574

The Github Action then should download the integration test files as [documented](https://github.com/hasenbanck/matroska-demuxer?tab=readme-ov-file#integration-test). It would be best if the Action also caches those files.

hasenbanck commented:
>Would you be interested to add functionality to be able to read tracks content of mkv in `matroska-demuxer` ?

Isn't that an orthogonal problem to demuxing? I think the reading and interpretation of for example subtitles is better handled in other crates, that could also be used for other containers, like MP4 or OGM.

Is there any functionality missing to support track content in this crate? If so, please add an issue for this outside of this PR for the CI pipeline.

gwen-lg committed:
ci: add cargo test management

include download test files from matroska project,
and cache the files to avoid dowload it each time workflow running.

gwen-lg commented:
PR updated with two fix for clippy lint, and test file downloading and caching :+1: .

I agreed than interpretation is not the purpose of `matroska-demuxer`.
For reading `matroska-demuxer` is already capable ? I haven't investigate much on this.

hasenbanck commented:
CI now looks fine now. I will merge it and close this PR. Please open another issue, if you think we need to support other features.
